source/common/upstream/outlier_detection_impl.cc

```diff
@@ -282,29 +282,41 @@ DetectorImpl::~DetectorImpl() {
   }
 }
 
-std::shared_ptr<DetectorImpl> DetectorImpl::create(
-    const Cluster& cluster, const envoy::config::cluster::v3::OutlierDetection& config,
-    Event::Dispatcher& dispatcher, Runtime::Loader& runtime, TimeSource& time_source,
-    EventLoggerSharedPtr event_logger, Random::RandomGenerator& random) {
+std::shared_ptr<DetectorImpl>
+DetectorImpl::create(Cluster& cluster, const envoy::config::cluster::v3::OutlierDetection& config,
+                     Event::Dispatcher& dispatcher, Runtime::Loader& runtime,
+                     TimeSource& time_source, EventLoggerSharedPtr event_logger,
+                     Random::RandomGenerator& random) {
   std::shared_ptr<DetectorImpl> detector(
       new DetectorImpl(cluster, config, dispatcher, runtime, time_source, event_logger, random));
 
   if (detector->config().maxEjectionTimeMs() < detector->config().baseEjectionTimeMs()) {
     throw EnvoyException(
         "outlier detector's max_ejection_time cannot be smaller than base_ejection_time");
   }
-
   detector->initialize(cluster);
 
   return detector;
 }
 
-void DetectorImpl::initialize(const Cluster& cluster) {
+void DetectorImpl::initialize(Cluster& cluster) {
   for (auto& host_set : cluster.prioritySet().hostSetsPerPriority()) {
     for (const HostSharedPtr& host : host_set->hosts()) {
       addHostMonitor(host);
     }
   }
+
+  if (cluster.healthChecker() != nullptr) {
+    cluster.healthChecker()->addHostCheckCompleteCb(
+        [this](HostSharedPtr host, HealthTransition changed_state) {
+          if (Runtime::runtimeFeatureEnabled(
+                  "envoy.reloadable_features.successful_active_health_check_uneject_host") &&
+              changed_state == HealthTransition::Changed &&
+              host->coarseHealth() == Host::Health::Healthy) {
+            unejectHost(host);
```

Comment:
Seems like a consequence of this is that not only do we uneject the host on hc success, we also reset any outlier detector counters etc on a successful hc (when the host was previously unhealthy)? This seems in line with considering active hc as an authoritative check, but seems like a bigger change than I first anticipated. We probably want this behavior documented in the outlier detention docs and reflected in the change log

Reply:
Makes sense. Updated